### PR TITLE
Update env to include delete-im, remove-EnvVal changes

### PR DESCRIPTION
**Pull request by jayz22**
### What

Update env to include delete-im, remove-EnvVal changes

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: soroban-sdk/src/accounts.rs [hunk 1]
```diff
@@ -111,16 +111,30 @@ impl PartialOrd for AccountId {
...
+            .0
+            .env
+            .obj_cmp(self.0.obj.to_raw(), other.0.obj.to_raw());
+        if v == 0 {
+            Ordering::Equal
+        } else if v < 0 {
+            Ordering::Less
+        } else {
+            Ordering::Greater
+        }
```
Comment (leighmcculloch) [inline soroban-sdk/src/accounts.rs:125]: I believe this is the same thing:
```suggestion
        v.cmp(&0)
```

File: soroban-sdk/src/accounts.rs [hunk 2]
```diff
@@ -421,7 +435,11 @@ impl testutils::Accounts for Accounts {
                     last_modified_ledger_seq: 0,
                     ext: xdr::LedgerEntryExt::V0,
                 };
-                storage.put(&k, &v)
+                storage.put(
+                    &k,
+                    &v,
+                    soroban_env_host::budget::AsBudget::as_budget(env.host()),
```
Comment (leighmcculloch) [inline soroban-sdk/src/accounts.rs:441]: What does as budget do in these calls now?

File: soroban-sdk/src/env.rs [hunk 3]
```diff
@@ -49,8 +50,17 @@ pub use internal::TryFromVal;
...
+pub struct EnvVal {
+    pub env: Env,
+    pub val: RawVal,
+}
+
+#[derive(Clone)]
+pub struct EnvObj {
+    pub env: Env,
+    pub obj: Object,
+}
```
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:63]: I think we can get rid of these types, at least that's what I imagined we'd do, and just put Env and Object inside each types like Bytes, Vec, etc. This is fine too though.

**Review by leighmcculloch — APPROVED** [review]
Small suggestions inline, otherwise lgtm.

I would consider getting rid of the EnvVal and EnvObj types. The goal was in part to get rid of unnecessary concepts, but this PR seems to attempt to maintain those two concepts, which I think are unnecessary and I regret adding them. Because they just make one more thing to learn when navigating this code.

**Comment by jayz22** [comment]
> I would consider getting rid of the EnvVal and EnvObj types. 

I would love so as well. I already removed EnvVal.
However, when I replace EnvObj (replacing with `Env` + `Object`) for `Map` and `Vec`: 
`pub struct Map<K, V>(Env, Object, PhantomData<K>, PhantomData<V>);`
`repr(transparent)` was giving me problem "transparent struct needs at most one non-zero-sized field, but has 2". Is it ok that we remove `repr(transparent)`?

**Comment by leighmcculloch** [comment]
> Is it ok that we remove `repr(transparent)`?

I think yes. The attributes guarantees that the ABI of the inner value and struct are the same, so that you could transmute between them. More details [here](https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent). We only need it on types that cross extern/exported functions, and the only types we have that do that are `RawVal` and `Object`. So it is critical that we keep it on those types, but other types no.

File: soroban-sdk/src/accounts.rs (hunk 2 above)
Comment (jayz22) [reply soroban-sdk/src/accounts.rs:434]: Logically same as before. It's just now a `Budget` is explicitly passed in every storage call, whereas previously the `Storage` owns a `MeteredOrdMap` which owns a `Budget` reference (same one as the host owns). Internally it charges the budget before preforming the corresponding map operation.

File: soroban-sdk/src/env.rs (hunk 3 above)
Comment (jayz22) [reply soroban-sdk/src/env.rs]: I have removed both of them.

**Review by leighmcculloch — APPROVED** [review]
🎉 👏🏻

**Pull request merged**